sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -235,9 +236,6 @@ object JavaTypeInference {
           path :: Nil,
           returnNullable = false)
 
-      case c if c == classOf[java.lang.String] =>
-        Invoke(path, "toString", ObjectType(classOf[String]))
```

Comment:
The `ScalaReflection` does the same thing, do we have a problem there too?

AFAIK the `path` should be a string type column, and it's always safe to call `UTF8String.toString`. My gut feeling is, we miss to add `Upcast` somewhere in `JavaTypeInference`.

Reply:
> AFAIK the `path` should be a string type column

The sample code tried to bind IntegerType column to String field in Java bean, which looks to break your expectation. (I guess ScalaReflection would not encounter this case.) 

Spark doesn't throw error for this case - actually Spark would show undefined behavior, compilation failures on generated code, even might be possible to throw runtime exception.

Reply:
> The sample code tried to bind IntegerType column to String field in Java bean

In scala, we can also do this and Spark will add `Upcast`. e.g. `spark.range(1).as[String].collect` works fine.

I did a quick search and `JavaTypeInference` has no `Upcast`. We should fix it and follow `ScalaReflection`

Reply:
Ah OK. I'll check and address it. Maybe it would be a separate PR if it doesn't fix the new test.

Reply:
Yeah your suggestion seems to work nicely! I left comment to ask which approach to choose: please compare both approach and comment. Thanks!